### ISSUE TITLE
.index.diff fails and/or doesn't list any change

### ISSUE THREAD

**Issue by yarikoptic** [gitpython-developers/GitPython#379]
So I have a local test repo which has staged changes, some non-staged changes, and untracked file (I am on branch b1):

```
$> git status
On branch b1
Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)

	modified:   123452

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   111

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	untracked
```

trying now to get this information (GitPython 1.0.1-108-g7b85487) (what differences were staged, what not) following docs on http://gitpython.readthedocs.org/en/stable/tutorial.html#obtaining-diff-information

```
$> PYTHONPATH=/home/yoh/deb/gits/python-git python -c 'import git; repo = git.Repo(".");  print repo.head, repo.index; print repo.index.diff(repo.head)' 
HEAD <git.index.base.IndexFile object at 0x7f3d482a6100>
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/yoh/deb/gits/python-git/git/index/util.py", line 69, in check_default_index
    return func(self, *args, **kwargs)
  File "/home/yoh/deb/gits/python-git/git/index/base.py", line 1207, in diff
    raise ValueError("other must be None, Diffable.Index, a Tree or Commit, was %r" % other)
ValueError: other must be None, Diffable.Index, a Tree or Commit, was <git.HEAD "HEAD">
```
and
```
$> PYTHONPATH=/home/yoh/deb/gits/python-git python -c 'import git; repo = git.Repo(".");  print repo.index.diff()'         
[]
```
... so am I mistreating GitPython and there is some other way to get list of those files, or a it is a bug?

**Comment by yarikoptic** [comment]
d'oh -- I read "None" and spell as nothing ;)  so for difference to work tree that works.  For difference with head -- need to use `head.commit`, so I guess either docs should be adjusted or may be better it digests HEAD and takes its .commit.  Cheers.

**Comment by yarikoptic** [comment]
btw -- what is then intended to be output of a diff to nothing? (i.e. when no *args)?

**Comment by yarikoptic** [comment]
related:  workaround `r.index.diff(r.head.commit)` doesn't work in cases when it is the first commit to be committed in the branch, so there is no commit associated with the head yet.
```
...
  File "/home/yoh/proj/datalad/datalad/datalad/crawler/nodes/annex.py", line 601, in _commit_versions
    staged = process_diff(repo.index.diff(repo.head.commit))
  File "/usr/lib/python2.7/dist-packages/git/refs/symbolic.py", line 183, in _get_commit
    obj = self._get_object()
  File "/usr/lib/python2.7/dist-packages/git/refs/symbolic.py", line 176, in _get_object
    return Object.new_from_sha(self.repo, hex_to_bin(self.dereference_recursive(self.repo, self.path)))
  File "/usr/lib/python2.7/dist-packages/git/refs/symbolic.py", line 125, in dereference_recursive
    hexsha, ref_path = cls._get_ref_info(repo, ref_path)
  File "/usr/lib/python2.7/dist-packages/git/refs/symbolic.py", line 157, in _get_ref_info
    raise ValueError("Reference at %r does not exist" % ref_path)
ValueError: Reference at 'refs/heads/incoming' does not exist
```

so what is the way to obtain information on what was staged so far in such a case?

**Comment by Byron** [comment]
Referring to the comment above only: Please have a look at [this reply on stackoverflow](http://stackoverflow.com/questions/33916648/get-the-diff-details-of-first-commit-in-gitpython) for a way on how to handle this.
Git itself handles this case, GitPython unfortunately does not.

**Comment by Byron** [comment]
About the issue: I believe it's more of a misunderstanding on which inputs are valid, and what they are doing. The linked documentation is supposed to be recent and working.
With the information on how to handle the first commit with diffs, you should be setup to make this work. Please close this issue if you think similarly.
Thank you

**Comment by yarikoptic** [comment]
yikes -- didn't know about the hexsha for empty tree ;)  it would be nice if GitPython took natively advantage of that knowledge instead of everyone hardcoding that one in their code

as the issue/docs:  `repo.index.diff(repo.head)` listed in http://gitpython.readthedocs.org/en/stable/tutorial.html#obtaining-diff-information is not working -- that is what I have reported.

**Comment by Byron** [comment]
Of course it would be nicer. If you find a good spot where to put that code, I will be happy to merge a related PR.

Thanks for clarifying the issue ! What I don't understand is how it doesn't work, as the code you see there is taken [straight from a test](https://github.com/gitpython-developers/GitPython/blob/84869674124aa5da988188675c1336697c5bcf81/git/test/test_docs.py#L394-L403).
What is the expected vs the observed behaviour - just so that we are absolutely clear about what need to be changed.
Thanks again.

**Comment by yarikoptic** [comment]
re head: nope -- in the test you don't have "repo.head" -- you have "HEAD" ;)

re good place to stick -- may be right there whenever HEAD is not defined? e.g.

```
$> PYTHONPATH=/home/yoh/deb/gits/python-git python -c 'import git; repo = git.Repo(".");  print repo.head.commit'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/yoh/deb/gits/python-git/git/refs/symbolic.py", line 183, in _get_commit
    obj = self._get_object()
  File "/home/yoh/deb/gits/python-git/git/refs/symbolic.py", line 176, in _get_object
    return Object.new_from_sha(self.repo, hex_to_bin(self.dereference_recursive(self.repo, self.path)))
  File "/home/yoh/deb/gits/python-git/git/refs/symbolic.py", line 125, in dereference_recursive
    hexsha, ref_path = cls._get_ref_info(repo, ref_path)
  File "/home/yoh/deb/gits/python-git/git/refs/symbolic.py", line 157, in _get_ref_info
    raise ValueError("Reference at %r does not exist" % ref_path)
ValueError: Reference at 'refs/heads/master' does not exist
```

**Comment by Byron** [comment]
Thank you. The situation presented here seems to be a repository without any commit, i.e. one that was just initialized. GitPython doesn't handle these situations automatically at all, thus the behaviour would have to be considered 'normal', even though it's annoying for every client out there having to handling this specifically if such a situation can arise.

As I don't see a bug here, I am not sure how else I can help.

**Comment by yarikoptic** [comment]
uff... I guess it is my fault since I have somewhat melded two issues into one.

1. `repo.index.diff(repo.head)` although listed in documentation it blows -- I have examples above and that has nothing to do with 2.  Either documentation needs to be adjusted or code+tests to support obvious `repo.head` in addition to `"HEAD"`

2. operation on first yet to be committed tree.  If GitPython aims to support what git natively supports, then it should be handled.  If not -- then I guess "not" and would be nice if at least there would be a note on that in docs

**Comment by Byron** [comment]
About `1)`: I don't find `repo.head` in the context of diff anywhere in the [docs](https://gitpython.readthedocs.org/en/stable/tutorial.html). My argument is that all examples are [executed ](https://github.com/gitpython-developers/GitPython/blob/ad3931357e5bb01941b50482b4b53934c0b715e3/git/test/test_docs.py), and verified to be working.

About `2)`: Even though it would be nice to handle this gracefully, it will only be implemented if someone submits a PR. Considering you ran into this issue, you could put the note about this limitation where you would have wanted to find it, and submit a PR.

**Comment by yarikoptic** [comment]
1) indeed was fixed in eb6beb75aaa269a1e7751d389c0826646878e5fc awhile back but I guess either I was accessing cached content or website wasn't updated only until recently.  now all kosher

2) fair enough.  I might get to doing the PR thing ;)

**Comment by Byron** [comment]
Closing this one in favour of the linked issue.
